hypothesis-python/docs/how-to/type-strategies.rst

```diff
@@ -0,0 +1,57 @@
+Write type hints for strategies
+===============================
+
+Hypothesis provides type hints for all standard strategies:
+
+.. code-block:: python
+
+    from hypothesis import strategies as st
+
+    # SearchStrategy[int]
+    reveal_type(st.integers())
+
+    # SearchStrategy[list[int]]
+    reveal_type(st.lists(st.integers()))
+
+|SearchStrategy| is the type of a strategy. You can use it to write type hints for your own strategies:
+
+.. code-block:: python
+
+    from hypothesis import strategies as st
+    from hypothesis.strategies import SearchStrategy
+
+
+    @st.composite
+    def all_ints_or_floats(draw) -> SearchStrategy[int] | SearchStrategy[float]:
+        all_integers = draw(st.booleans())
+        return st.integers() if all_integers else st.floats()
```

Comment:
This seems like a pretty unnatural use-case: `@st.composite` wraps the returned value, so we're creating a strategy which generates a strategy!  For `@st.composite`, users should typically annotate their function as returning whatever value (e.g. `int | float`), and let the decorator wrap it.

I'd probably make this example use the `json()` strategy helper we used for the tutorial, and explicitly note afterwards that composite strategies don't need `SearchStrategy` in their annotations.

Reply:
Typing the return value of examples from strategies, like `int | float`, is relatively boring, so I did actually want an example where we used `SearchStrategy` parametrized over a type. It's an unusual construction (and I don't claim this example is good), but that's why it's a howto!

Though now that you say this, it's a good idea to add a separate example just for `@st.composite`, showing that you annotate the example types, not `SearchStrategy`.

Reply:
Agree that we want a `-> SearchStrategy[]` example, I just think that function shouldn't have `@st.composite`.